control/velocity_controller/src/LQR_setup.cpp

```diff
@@ -0,0 +1,356 @@
+        0, -1, 0, 0, 0;
+    augmented_input_matrix << inertia_matrix_inv(0, 0),
+        inertia_matrix_inv(0, 1), inertia_matrix_inv(0, 2), 0, 0, 0,
+        inertia_matrix_inv(1, 0), inertia_matrix_inv(1, 1),
+        inertia_matrix_inv(1, 2), 0, 0, 0, inertia_matrix_inv(2, 0),
+        inertia_matrix_inv(2, 1), inertia_matrix_inv(2, 2), 0, 0, 0;
+    {
+        std::ostringstream oss;
+        oss << augmented_system_matrix.format(fmt);
+        std::string A_str = oss.str();
+        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "A:\n%s", A_str.c_str());
+    }
+    return;
+};
+Eigen::Vector<double, 6> LQRController::update_error(
+    Guidance_data guidance_values,
+    State states) {
+    double surge_error = guidance_values.surge - states.surge;
+    double pitch_error = ssa(guidance_values.pitch - states.pitch);
+    double yaw_error = ssa(guidance_values.yaw - states.yaw);
+
+    integral_error_surge =
+        anti_windup(i_surge, surge_error, integral_error_surge, surge_windup);
+    integral_error_pitch =
+        anti_windup(i_pitch, pitch_error, integral_error_pitch, pitch_windup);
+    integral_error_yaw =
+        anti_windup(i_yaw, yaw_error, integral_error_yaw, yaw_windup);
+
+    Eigen::Vector<double, 6> state_error = {
+        -surge_error,         -pitch_error,         -yaw_error,
+        integral_error_surge, integral_error_pitch, integral_error_yaw};
+    return state_error;
+}
+Eigen::Vector<double, 3> LQRController::saturate_input(
+    Eigen::Vector<double, 3> u) {
+    double force_x, torque_y, torque_z;
+    std::tie(surge_windup, force_x) = saturate(u[0], surge_windup, max_force);
+    std::tie(pitch_windup, torque_y) = saturate(u[1], pitch_windup, max_force);
+    std::tie(yaw_windup, torque_z) = saturate(u[2], yaw_windup, max_force);
+    return {force_x, torque_y, torque_z};
+}
+Eigen::Vector<double, 3> LQRController::calculate_lqr_u(
+    State state,
+    Guidance_data guidance_values) {
+    update_augmented_matrices(calculate_coriolis_matrix(
+        state.pitch_rate, state.yaw_rate, state.sway_vel, state.heave_vel));
+    LQRsolveResult result =
+        solve_k_p(augmented_system_matrix, augmented_input_matrix,
+                  state_weight_matrix, input_weight_matrix);
+    if (result.INFO != 0) {
+        return {9999, 9999, 9999};  // Need to fix
+    }
+    Eigen::Matrix<double, 6, 1> state_error =
+        update_error(guidance_values, state);
+    Eigen::Vector<double, 3> u = saturate_input(-(result.K * state_error));
+    return u;
+}
+void LQRController::reset_controller() {
+    integral_error_surge = 0.0;
+    integral_error_pitch = 0.0;
+    integral_error_yaw = 0.0;
+
+    surge_windup = false;
+    pitch_windup = false;
+    yaw_windup = false;
+    return;
+}
+
+extern "C" {
+// Fortran subroutine for solving symplectic Schur decomposition(double
+// precision version)
+void sb02mt_(const char* JOBG,
+             const char* JOBL,
```

Comment:
Genuine curiosity, where did this come from? (If you still want to keep the online solver, Eigen supports LDLT out-of-the-box. Alternatively, you could make a separate solver that runs entirely offline or just once during construction?)